frame/gift/src/lib.rs

```diff
@@ -0,0 +1,270 @@
+
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// The Balances of your system.
+		type Currency: ReservableCurrency<Self::AccountId>;
+		/// The additional deposit needed to place a gift. Should be greater than the existential
+		/// deposit to avoid killing the gifter account.
+		type GiftDeposit: Get<BalanceOf<Self>>;
+		/// The minimum gift amount. Should be greater than the existential deposit.
+		type MinimumGift: Get<BalanceOf<Self>>;
+		/// Pallet hook for when a gift is claimed.
+		type OnGift: OnTransfer<Self::AccountId, BalanceOf<Self>>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[derive(Encode, Decode)]
+	pub struct GiftInfo<AccountId, Balance> {
+		gifter: AccountId,
+		deposit: Balance,
+		amount: Balance,
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn gifts)]
+	pub type Gifts<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		T::AccountId, // Gift Controller
+		GiftInfo<T::AccountId, BalanceOf<T>>,
+	>;
+
+	#[pallet::event]
+	#[pallet::metadata(T::AccountId = "AccountId", BalanceOf<T> = "Balance")]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A gift has been created! [amount, controller]
+		GiftCreated(BalanceOf<T>, T::AccountId),
+		/// A gift has been claimed! [controller, amount, to]
+		GiftClaimed(T::AccountId, BalanceOf<T>, T::AccountId),
+		/// A gift has been removed :( [to]
+		GiftRemoved(T::AccountId),
+	}
+
+	// Errors inform users that something went wrong.
+	#[pallet::error]
+	pub enum Error<T> {
+		/// User already has a pending gift.
+		PendingGift,
+		/// Don't be cheap... Get your friend a good gift!
+		GiftTooSmall,
+		/// An overflow has occurred.
+		Overflow,
+		/// A gift does not exist for this user.
+		NoGift,
+		/// You are not the owner of this gift.
+		NotGifter,
+	}
+
+	// Dispatchable functions allows users to interact with the pallet and invoke state changes.
+	// These functions materialize as "extrinsics", which are often compared to transactions.
+	// Dispatchable functions must be annotated with a weight and must return a DispatchResult.
+	#[pallet::call]
+	impl<T:Config> Pallet<T> {
+		/// Create a gift.
+		///
+		/// A gift is some balance which is reserved from the calling account, and placed under
+		/// control of the `controller` account.
+		///
+		/// The credentials of this controller account can be shared with anyone, allow them to then
+		/// transfer the balance to a final account, without needing to expose the final accounts
+		/// credentials. This is done via the `claim` extrinsic.
+		///
+		/// NOTE: The fee of this extrinsic should take into account the full cost of `claim`, which
```

Comment:
but the claim will be processed by a different block producer and this is just unfair. an economically rational block producer will exclude all `claim` transaction in hope to pack more transactions with fees.